ls1intum/Artemis
# `Exam mode`: Add basic structure for test exams and refactor exam creation dialog

### Checklist
#### General
#### Server
#### Client

### Motivation 
The current Exam Mode of Artemis is designed to facilitate end-of-semester exams, which have a fixed working time and date. 
In preparation of the end-of-semester exams, students often use legacy exams or test exams provided by the course to practice the content. Using the existing Exam Mode of Artemis to conduct such test exams has multiple drawbacks. For example, the working time is fixed to a predefined time slot, where students must conduct the exams. Multiple conductions are also not possible. That's why this PR introduces TestExams as a new Exam Mode, wich will be fully implemented with further PRs. This PR focuses on the creation of Exams as well as the Entity changes.

### Description
**Database and Entity changes**
The class `Exam` is extended with the attributes `isTestExam : boolen` and `workingTime: Integer`. The first attribute serves as a flag for a TestExam (-> next paragraph), the second attribute stores the exam working time in seconds. 
Using Liquibase-Changelog the workingTime is calculated for all previous exams.

**Exam Creation Page**
The Exam Creation Page was restructured into four parts to improve the lucidity of the page:
-	Exam Configuration
-	Exam Conduction
-	Exam Exercises
-	Exam Assessment and Student Review
-	Exam Texts

Tooltips have been added to the input boxes in the restructured section *Exam Conduction* to provide a better user guidance.  

**Introducing Test Exams**
Test exams aim to give students the opportunity to conduct an exam several times within the specified working window. Students receive automated feedback for each run.
In the first step, the possibility to create a test exam is implemented. For this purpose, a new `exam-mode-picker` has been added, which allows to choose between the modes Exam *(internally RealExam)* and TestExam during the exam creation. *This feature is currently only accessible for Admins, as the Exam Conduction is missing*
For RealExams, the working time is automatically calculated based on the start- and end-date. For TestExams, the user must input a working time.
As test exams are automatically assessed, the "Exam Assessment and Student Review" section is hidden.

### Steps for testing
Log in with an **Admin**-account
1. Open the dialog to create a new exam
2. Make sure that the picker-component correctly switches between the two modes
3. Make sure that all input boxes in the section *Exam Conduction* and all tooltips have meaningful labels and hints (in both languages).

Try to create an exam:
1. The working time should be automatically calculated from the Start- and End-Date, the User should not be able to set the working time himself
2.  Store the exam to the database and reopen the exam from the Exam Management to make sure, that all inputs are correctly stored in the database and send to the client
3. Make sure, that the exam mode cannot be changed after creation.

Try to create an TestExam:
1. The working time should be a required input from the user
4. The section “Exam Assessment and Student Review” should be hidden
5. Store the TestExam to the database and reopen the TestExam from the Exam Management to make sure, that all inputs are correctly stored in the database and send to the client
6. Login with a non-admin account and make sure, that the user cannot choose between the Exam and TestExam-Mode (as the TestExams are currently WIP and the backend for conducting the exam is missing) 

Try to find (hopefully no) bugs:
1. Leave out values and try to save the exam
2. Input values outside of a rational range

### Review Progress

#### Code Review
#### Manual Tests
 
### Test Coverage

| Class/File | Branch | Line |
--- | --- | ---
| ExamRessource.java | 98% | 90% |
| [exam-update.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS3227-24/artifact/shared/Coverage-Report-Client-Tests/app/exam/manage/exams/index.html) | 88% | 94% |
 
### Screenshots
![image](https://user-images.githubusercontent.com/94070506/161426657-fa57ce46-75b2-44ce-acf1-d70e137154af.png)
